### DISCUSSION TITLE
The process of replacing the preinstalled OS strongSwan with a new build from source code

### DISCUSSION THREAD

**Discussion by JimChen111** [strongswan/strongswan#2315] [Q&A]
Hi,

I'm trying to replace the preinstalled strongSwan on the OS with a new version built from source code. I currently can build and `make install` the package from the source code. However, after the install, I still can't enable the strongSwan with `systemctl` which complains about `Unit strongswan.service not found`.

Here are my steps:
1) Uninstall the current strongSwan on my OS
```
systemctl stop strongswan
sudo apt purge strongswan*
```

2) Make a new strongSwan from source code. My cmd:
```
./configure --prefix=/usr --sysconfdir=/etc \
--enable-openssl \
--enable-systemd \
--with-systemdsystemunitdir \
--with-systemduserunitdir \
--enable-swanctl \
--disable-aes \
--disable-hmac \
--disable-gmp \
--disable-des \
--disable-rc2 \
--disable-sha2 \
--disable-sha1 \
--disable-md5 \
--disable-random \
--disable-nonce \
--disable-x509 \
--disable-revocation \
--disable-constraints \
--disable-pubkey \
--disable-pkcs1 \
--disable-pkcs7 \
--disable-pkcs12 \
--disable-pgp \
--disable-sshkey \
--disable-pem \
--disable-pkcs8 \
--disable-fips-prf \
--disable-xcbc \
--disable-cmac \
--disable-kdf \
--disable-drbg 
```
```
make
make install
```

3. Re-enable the strongSwan service with `systemctl`
`systemctl start strongswan`

However, currently, when I try to run step 3, there is an error complaining `Failed to start strongswan.service: Unit strongswan.service not found.`

Any steps above may have problems? Thanks in advance!

**Comment by tobiasbrunner** [comment]
> --with-systemdsystemunitdir \

This option takes a path. Not specifying one won't work (it sets it to `yes`, so the systemd unit will be installed in a weird directory, probably `$DESTDIR/yes`). However, it's usually also not necessary to specify it explicitly as it's determined automatically using `pkgconfig`.

> --with-systemduserunitdir \

That's not an option for the configure script.

> sudo apt purge strongswan*

Note that not all Debian/Ubuntu packages related to strongSwan start with that prefix. Some start with `charon`, others with `libcharon` or `libstrongswan`.

**Reply by JimChen111** [reply]
Thanks Tobias.

I removed the two `--with-system*` flags. Now when I try to start strongSwan by `systemctl`, it reports:
```
systemctl start strongswan
Job for strongswan.service failed because the control process exited with error code.
See "systemctl status strongswan.service" and "journalctl -xe" for details.
```

`systemctl status strongswan.service` gives me:
```
systemctl status strongswan.service
● strongswan.service - strongSwan IPsec IKEv1/IKEv2 daemon using swanctl
     Loaded: loaded (/lib/systemd/system/strongswan.service; disabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Mon 2024-07-01 20:25:05 UTC; 14s ago
    Process: 2783261 ExecStart=/usr/sbin/charon-systemd (code=exited, status=66)
   Main PID: 2783261 (code=exited, status=66)
     Status: "charon initialization failed"

Jul 01 20:25:05 gpc-adhoc-665c10bevm-worker-node0-zone1 charon-systemd[2783261]: loading ca certificates from '/etc/ipsec.d/cacerts'
Jul 01 20:25:05 gpc-adhoc-665c10bevm-worker-node0-zone1 charon-systemd[2783261]: loading aa certificates from '/etc/ipsec.d/aacerts'
Jul 01 20:25:05 gpc-adhoc-665c10bevm-worker-node0-zone1 charon-systemd[2783261]: loading ocsp signer certificates from '/etc/ipsec.d/ocspcerts'
Jul 01 20:25:05 gpc-adhoc-665c10bevm-worker-node0-zone1 charon-systemd[2783261]: loading attribute certificates from '/etc/ipsec.d/acerts'
Jul 01 20:25:05 gpc-adhoc-665c10bevm-worker-node0-zone1 charon-systemd[2783261]: loading crls from '/etc/ipsec.d/crls'
Jul 01 20:25:05 gpc-adhoc-665c10bevm-worker-node0-zone1 charon-systemd[2783261]: loading secrets from '/etc/ipsec.secrets'
Jul 01 20:25:05 gpc-adhoc-665c10bevm-worker-node0-zone1 charon-systemd[2783261]: failed to load 1 critical plugin feature
Jul 01 20:25:05 gpc-adhoc-665c10bevm-worker-node0-zone1 systemd[1]: strongswan.service: Main process exited, code=exited, status=66/NOINPUT
Jul 01 20:25:05 gpc-adhoc-665c10bevm-worker-node0-zone1 systemd[1]: strongswan.service: Failed with result 'exit-code'.
Jul 01 20:25:05 gpc-adhoc-665c10bevm-worker-node0-zone1 systemd[1]: Failed to start strongSwan IPsec IKEv1/IKEv2 daemon using swanctl.
```

It seems it mentions `failed to load 1 critical plugin feature`. Do you know which one(s) I'm missing?

**Reply by tobiasbrunner** [reply]
I don't, please read the complete log (e.g. via `journalctl -eu strongswan`).
